File: app/javascript/src/shared/policies/product_category_policy.js
import PolicyBase from './policy_base'

export default class ProductCategoryPolicy extends PolicyBase {
  constructor(user, record = null) {
    super(user, record)
  }

  index() {
    return true
  }

  show() {
    return true
  }

  create() {
    return this.forAdminOwnerAndSuperAndStockManagerOnly()
  }

  update() {
    return this.forAdminOwnerAndSuperAndStockManagerOnly()
  }

  // helpers

  forAdminOwnerAndSuperAndStockManagerOnly() {
    if (this.isAdmin()) {
      return this.user.isOwner() || this.user.isSuper() || this.user.isStockManager()
    } else {
      return false
    }
  }
}

File: app/policies/product_category_policy.rb
class ProductCategoryPolicy < ApplicationPolicy
  def index?
    true
  end

  def show?
    index?
  end

  def create?
    return user.super? || user.owner? || user.stock_manager? if admin?

    false
  end

  def update?
    create?
  end

  class Scope < Scope
    def resolve
      scope
    end
  end
end
